library/Network/CloudSeeder/Main.hs

```diff
@@ -98,53 +125,131 @@ runAppM (AppM x) = do
+decodeTemplate :: (AsCliError e, MonadError e m) => T.Text -> m Template
+decodeTemplate templateBody = do
+  let decodeOrFailure = decodeEither (encodeUtf8 templateBody) :: Either String Template
+  either (throwing _CliTemplateDecodeFail) return decodeOrFailure
 
-  templateBody <- readFile $ nameToDeploy <> ".yaml"
+getOutputs :: (AsCliError e, MonadError e m, MonadCloud m, Traversable t) => t T.Text -> T.Text -> T.Text -> m (S.Set (T.Text, T.Text))
+getOutputs dependencies env appName = do 
+  maybeOutputs <- mapM (\stackName -> (stackName,) <$> getStackOutputs (mkFullStackName env appName stackName)) dependencies
+  let outputsOrFailure = runErrors $ traverse (extractResult (flip const)) maybeOutputs
+  either (throwing _CliMissingDependencyStacks) (return . S.fromList . concatMap M.toList) outputsOrFailure
 
-  let decodeOrFailure = decodeEither (encodeUtf8 templateBody) :: Either String Value
-  template <- either (throwing _CliTemplateDecodeFail) return decodeOrFailure
-  let requiredParams = template ^. key "Parameters" . _Object.to keys
+mkFullStackName :: T.Text -> T.Text -> T.Text -> StackName
+mkFullStackName env appName stackName = StackName $ env <> "-" <> appName <> "-" <> stackName
 
-  maybeOutputs <- mapM (\stackName -> (stackName,) <$> getStackOutputs (mkStackName stackName)) dependencies
-  let outputsOrFailure = runErrors $ traverse (extractResult (flip const)) maybeOutputs
-  outputs <- either (throwing _CliMissingDependencyStacks) (return . concat) outputsOrFailure
+getTags :: (HasTagSet s1 (S.Set (T.Text, T.Text)), HasTagSet s (S.Set (T.Text, T.Text))) => s1 -> s -> T.Text -> T.Text -> S.Set (T.Text, T.Text)
+getTags config stackToDeploy env appName = baseTags <> globalTags <> localTags
+  where 
+    baseTags :: S.Set (T.Text, T.Text)
+    baseTags = [("cj:environment", env), ("cj:application", appName)]
+    globalTags = config ^. tagSet
+    localTags = stackToDeploy ^. tagSet
+
+getPassedParameters :: ( MonadArguments e f, HasParameterSources s1 (S.Set (T.Text, ParameterSource)), HasParameterSources s (S.Set (T.Text, ParameterSource))) 
+                    => s1 -> s -> f (S.Set (T.Text, T.Text))
+getPassedParameters config stackToDeploy = do
+  let globalParamSources = config ^. parameterSources
+      localParamSources = stackToDeploy ^. parameterSources
+      isFlag (_, Flag) = True
+      isFlag (_, _) = False
+      paramFlags = S.fromList $ (globalParamSources <> localParamSources)^..folded.filtered isFlag
+      --TODO: use a more lensy approach, a la something like replacing "isFlag" w/ "(_2.is _Flag)"
+  S.fromList . M.toList <$> getOptions paramFlags
 
-  let globalParams = (config ^. parameters) ++ [("Env", env)]
+collectParameters :: (HasParameters s1 (S.Set (T.Text, T.Text)), HasParameters s (S.Set (T.Text, T.Text))) 
+                  => s1 -> s -> S.Set (T.Text, T.Text) -> T.Text -> S.Set (T.Text, T.Text) -> S.Set (T.Text, T.Text) -> S.Set (T.Text, T.Text)
```

Comment:
These type signatures are intense. 😳 They’re definitely awfully complex, too complex for me to visually scan. If we can’t simplify them, I think it would be good to at least add documentation comments for each of these top-level functions to give a brief description of what they do and how they fit into the overall process.